File: payment-core/go-services/internal/webhooks/remittance_webhook_service.go
package webhooks

import (
	"bytes"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"sync"
	"time"
)

type WebhookEvent struct {
	ID           string                 `json:"id"`
	RemittanceID string                 `json:"remittanceId"`
	Event        string                 `json:"event"`
	Data         map[string]interface{} `json:"data"`
	Timestamp    time.Time              `json:"timestamp"`
	Signature    string                 `json:"signature,omitempty"`
}

type WebhookDelivery struct {
	ID             string    `json:"id"`
	WebhookEventID string    `json:"webhookEventId"`
	URL            string    `json:"url"`
	Status         string    `json:"status"`
	Attempts       int       `json:"attempts"`
	LastAttemptAt  time.Time `json:"lastAttemptAt,omitempty"`
	NextRetryAt    time.Time `json:"nextRetryAt,omitempty"`
	ResponseCode   int       `json:"responseCode,omitempty"`
	ResponseBody   string    `json:"responseBody,omitempty"`
	Error          string    `json:"error,omitempty"`
}

type WebhookSubscription struct {
	ID        string    `json:"id"`
	UserID    string    `json:"userId"`
	URL       string    `json:"url"`
	Secret    string    `json:"secret"`
	Events    []string  `json:"events"`
	Active    bool      `json:"active"`
	CreatedAt time.Time `json:"createdAt"`
}

const (
	EventPaymentPending      = "payment.pending"
	EventPaymentConfirmed    = "payment.confirmed"
	EventPaymentFailed       = "payment.failed"
	EventConversionStarted   = "conversion.started"
	EventConversionCompleted = "conversion.completed"
	EventConversionFailed    = "conversion.failed"
	EventKYCInitiated        = "kyc.initiated"
	EventKYCApproved         = "kyc.approved"
	EventKYCRejected         = "kyc.rejected"
	EventAccountVerifying    = "account.verifying"
	EventAccountVerified     = "account.verified"
	EventAccountOpening      = "account.opening"
	EventAccountOpened       = "account.opened"
	EventTransferInitiated   = "transfer.initiated"
	EventTransferProcessing  = "transfer.processing"
	EventTransferCompleted   = "transfer.completed"
	EventTransferFailed      = "transfer.failed"
	EventRemittanceCreated   = "remittance.created"
	EventRemittanceCompleted = "remittance.completed"
	EventRemittanceFailed    = "remittance.failed"
	EventRemittanceCancelled = "remittance.cancelled"
)

type WebhookService struct {
	mu            sync.RWMutex
	subscriptions map[string]*WebhookSubscription
	events        map[string]*WebhookEvent
	deliveries    map[string]*WebhookDelivery
	httpClient    *http.Client
	maxRetries    int
	retryDelays   []int
	db            *sql.DB
}

func NewWebhookService() *WebhookService {
	return &WebhookService{
		subscriptions: make(map[string]*WebhookSubscription),
		events:        make(map[string]*WebhookEvent),
		deliveries:    make(map[string]*WebhookDelivery),
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
		maxRetries:  5,
		retryDelays: []int{60, 300, 900, 3600, 21600},
	}
}

func (s *WebhookService) CreateWebhookEvent(remittanceID, eventType string, data map[string]interface{}) (*WebhookEvent, error) {
	eventID := s.generateID("evt")

	event := &WebhookEvent{
		ID:           eventID,
		RemittanceID: remittanceID,
		Event:        eventType,
		Data:         data,
		Timestamp:    time.Now(),
	}

	s.mu.Lock()
	s.events[eventID] = event
	s.mu.Unlock()

	go s.persistEvent(event)
	go s.deliverWebhookEvent(event)

	return event, nil
}

func (s *WebhookService) deliverWebhookEvent(event *WebhookEvent) {
	subscriptions := s.getMatchingSubscriptions(event.Event)

	for _, sub := range subscriptions {
		delivery := &WebhookDelivery{
			ID:             s.generateID("del"),
			WebhookEventID: event.ID,
			URL:            sub.URL,
			Status:         "pending",
			Attempts:       0,
		}

		s.mu.Lock()
		s.deliveries[delivery.ID] = delivery
		s.mu.Unlock()

		go s.persistDelivery(delivery)
		s.attemptWebhookDelivery(event, delivery, sub)
	}
}

func (s *WebhookService) attemptWebhookDelivery(event *WebhookEvent, delivery *WebhookDelivery, subscription *WebhookSubscription) {
	delivery.Attempts++
	delivery.LastAttemptAt = time.Now()

	signature := s.generateWebhookSignature(event, subscription.Secret)

	payload := map[string]interface{}{
		"id":           event.ID,
		"event":        event.Event,
		"remittanceId": event.RemittanceID,
		"data":         event.Data,
		"timestamp":    event.Timestamp.Format(time.RFC3339),
	}

	jsonData, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", subscription.URL, bytes.NewBuffer(jsonData))
	if err != nil {
		delivery.Status = "failed"
		delivery.Error = fmt.Sprintf("Failed to create request: %v", err)
		s.scheduleRetry(delivery)
		return
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Webhook-Signature", signature)
	req.Header.Set("X-Webhook-Event", event.Event)
	req.Header.Set("X-Webhook-ID", event.ID)
	req.Header.Set("User-Agent", "PaymentSwitch-Webhooks/1.0")

	resp, err := s.httpClient.Do(req)
	if err != nil {
		delivery.Status = "failed"
		delivery.Error = fmt.Sprintf("Request failed: %v", err)
		s.scheduleRetry(delivery)
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	delivery.ResponseCode = resp.StatusCode
	delivery.ResponseBody = string(body)

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		delivery.Status = "delivered"
	} else {
		delivery.Status = "failed"
		delivery.Error = fmt.Sprintf("HTTP %d: %s", resp.StatusCode, resp.Status)
		s.scheduleRetry(delivery)
	}
}

func (s *WebhookService) scheduleRetry(delivery *WebhookDelivery) {
	if delivery.Attempts >= s.maxRetries {
		return
	}

	delayIndex := delivery.Attempts - 1
	if delayIndex >= len(s.retryDelays) {
		delayIndex = len(s.retryDelays) - 1
	}

	delay := s.retryDelays[delayIndex]
	delivery.NextRetryAt = time.Now().Add(time.Duration(delay) * time.Second)
	delivery.Status = "pending"
}

func (s *WebhookService) generateWebhookSignature(event *WebhookEvent, secret string) string {
	payload := map[string]interface{}{
		"id":           event.ID,
		"event":        event.Event,
		"remittanceId": event.RemittanceID,
		"data":         event.Data,
		"timestamp":    event.Timestamp.Format(time.RFC3339),
	}

	jsonData, _ := json.Marshal(payload)

	h := hmac.New(sha256.New, []byte(secret))
	h.Write(jsonData)
	return hex.EncodeToString(h.Sum(nil))
}

func (s *WebhookService) VerifyWebhookSignature(payload, signature, secret string) bool {
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(payload))
	expectedSignature := hex.EncodeToString(h.Sum(nil))

	return hmac.Equal([]byte(signature), []byte(expectedSignature))
}

func (s *WebhookService) getMatchingSubscriptions(eventType string) []*WebhookSubscription {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var matching []*WebhookSubscription
	for _, sub := range s.subscriptions {
		if sub.Active && s.matchesEventPattern(eventType, sub.Events) {
			matching = append(matching, sub)
		}
	}
	return matching
}

func (s *WebhookService) matchesEventPattern(event string, patterns []string) bool {
	for _, pattern := range patterns {
		regexPattern := regexp.QuoteMeta(pattern)
		regexPattern = regexp.MustCompile(`\\\*`).ReplaceAllString(regexPattern, ".+")
		regex := regexp.MustCompile("^" + regexPattern + "$")
		if regex.MatchString(event) {
			return true
		}
	}
	return false
}

func (s *WebhookService) CreateSubscription(userID, url string, events []string) (*WebhookSubscription, error) {
	secretBytes := make([]byte, 32)
	rand.Read(secretBytes)
	secret := hex.EncodeToString(secretBytes)

	subscription := &WebhookSubscription{
		ID:        s.generateID("sub"),
		UserID:    userID,
		URL:       url,
		Secret:    secret,
		Events:    events,
		Active:    true,
		CreatedAt: time.Now(),
	}

	s.mu.Lock()
	s.subscriptions[subscription.ID] = subscription
	s.mu.Unlock()

	go s.persistSubscription(subscription)
	return subscription, nil
}

func (s *WebhookService) UpdateSubscription(subscriptionID string, url *string, events []string, active *bool) (*WebhookSubscription, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	sub, exists := s.subscriptions[subscriptionID]
	if !exists {
		return nil, fmt.Errorf("subscription not found")
	}

	if url != nil {
		sub.URL = *url
	}
	if events != nil {
		sub.Events = events
	}
	if active != nil {
		sub.Active = *active
	}

	return sub, nil
}

func (s *WebhookService) DeleteSubscription(subscriptionID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.subscriptions[subscriptionID]; !exists {
		return fmt.Errorf("subscription not found")
	}

	delete(s.subscriptions, subscriptionID)
	go s.deleteSubscriptionFromDB(subscriptionID)
	return nil
}

func (s *WebhookService) GetWebhookDeliveries(eventID string) []*WebhookDelivery {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var deliveries []*WebhookDelivery
	for _, d := range s.deliveries {
		if d.WebhookEventID == eventID {
			deliveries = append(deliveries, d)
		}
	}
	return deliveries
}

func (s *WebhookService) RetryWebhookDelivery(deliveryID string) (*WebhookDelivery, error) {
	s.mu.Lock()
	delivery, exists := s.deliveries[deliveryID]
	if !exists {
		s.mu.Unlock()
		return nil, fmt.Errorf("delivery not found")
	}

	event, eventExists := s.events[delivery.WebhookEventID]
	s.mu.Unlock()

	if !eventExists {
		return nil, fmt.Errorf("event not found")
	}

	var subscription *WebhookSubscription
	s.mu.RLock()
	for _, sub := range s.subscriptions {
		if sub.URL == delivery.URL {
			subscription = sub
			break
		}
	}
	s.mu.RUnlock()

	if subscription == nil {
		return nil, fmt.Errorf("subscription not found")
	}

	s.attemptWebhookDelivery(event, delivery, subscription)
	return delivery, nil
}

func (s *WebhookService) ProcessPendingWebhooks() (processed, succeeded, failed int) {
	s.mu.RLock()
	var pendingDeliveries []*WebhookDelivery
	now := time.Now()

	for _, d := range s.deliveries {
		if d.Status == "pending" && (d.NextRetryAt.IsZero() || now.After(d.NextRetryAt)) {
			pendingDeliveries = append(pendingDeliveries, d)
		}
	}
	s.mu.RUnlock()

	for _, delivery := range pendingDeliveries {
		s.mu.RLock()
		event := s.events[delivery.WebhookEventID]
		var subscription *WebhookSubscription
		for _, sub := range s.subscriptions {
			if sub.URL == delivery.URL {
				subscription = sub
				break
			}
		}
		s.mu.RUnlock()

		if event == nil || subscription == nil {
			continue
		}

		s.attemptWebhookDelivery(event, delivery, subscription)
		processed++

		if delivery.Status == "delivered" {
			succeeded++
		} else {
			failed++
		}
	}

	return processed, succeeded, failed
}

func (s *WebhookService) GetWebhookEvent(eventID string) (*WebhookEvent, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	event, exists := s.events[eventID]
	if !exists {
		return nil, fmt.Errorf("event not found")
	}
	return event, nil
}

func (s *WebhookService) ListWebhookEvents(remittanceID string, limit, offset int) ([]*WebhookEvent, int) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var events []*WebhookEvent
	for _, e := range s.events {
		if e.RemittanceID == remittanceID {
			events = append(events, e)
		}
	}

	total := len(events)

	if offset >= len(events) {
		return []*WebhookEvent{}, total
	}

	end := offset + limit
	if end > len(events) {
		end = len(events)
	}

	return events[offset:end], total
}

func (s *WebhookService) TestWebhookEndpoint(url, secret string) (success bool, responseCode int, responseTime int64, err error) {
	startTime := time.Now()

	testEvent := &WebhookEvent{
		ID:           "evt_test",
		RemittanceID: "rem_test",
		Event:        "webhook.test",
		Data:         map[string]interface{}{"message": "This is a test webhook"},
		Timestamp:    time.Now(),
	}

	signature := s.generateWebhookSignature(testEvent, secret)

	payload := map[string]interface{}{
		"id":           testEvent.ID,
		"event":        testEvent.Event,
		"remittanceId": testEvent.RemittanceID,
		"data":         testEvent.Data,
		"timestamp":    testEvent.Timestamp.Format(time.RFC3339),
	}

	jsonData, _ := json.Marshal(payload)

	req, reqErr := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if reqErr != nil {
		return false, 0, time.Since(startTime).Milliseconds(), reqErr
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Webhook-Signature", signature)
	req.Header.Set("X-Webhook-Event", testEvent.Event)
	req.Header.Set("X-Webhook-ID", testEvent.ID)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, respErr := client.Do(req)
	responseTime = time.Since(startTime).Milliseconds()

	if respErr != nil {
		return false, 0, responseTime, respErr
	}
	defer resp.Body.Close()

	return resp.StatusCode >= 200 && resp.StatusCode < 300, resp.StatusCode, responseTime, nil
}

func (s *WebhookService) GetSubscription(subscriptionID string) (*WebhookSubscription, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	sub, exists := s.subscriptions[subscriptionID]
	if !exists {
		return nil, fmt.Errorf("subscription not found")
	}
	return sub, nil
}

func (s *WebhookService) GetUserSubscriptions(userID string) []*WebhookSubscription {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var subs []*WebhookSubscription
	for _, sub := range s.subscriptions {
		if sub.UserID == userID {
			subs = append(subs, sub)
		}
	}
	return subs
}

func (s *WebhookService) generateID(prefix string) string {
	bytes := make([]byte, 16)
	rand.Read(bytes)
	return fmt.Sprintf("%s_%s", prefix, hex.EncodeToString(bytes))
}

File: server/services/remittanceWebhookService.ts
/**
 * Remittance Webhook Service
 * 
 * Handles webhook delivery for remittance events with:
 * - Signature verification (HMAC-SHA256)
 * - Automatic retries with exponential backoff
 * - Delivery tracking and logging
 * - Event filtering and subscriptions
 */

import crypto from 'crypto';
import { createChildLogger } from '../lib/logger';
import { getDb } from '../db';
import { eq, desc, and, lte } from 'drizzle-orm';
import { webhookDeliveries } from '../../drizzle/payments-schema';

const log = createChildLogger('webhooks');

export interface WebhookEvent {
  id: string;
  remittanceId: string;
  event: string;
  data: Record<string, any>;
  timestamp: Date;
  signature?: string;
}

export interface WebhookDelivery {
  id: string;
  webhookEventId: string;
  url: string;
  status: 'pending' | 'delivered' | 'failed';
  attempts: number;
  lastAttemptAt?: Date;
  nextRetryAt?: Date;
  responseCode?: number;
  responseBody?: string;
  error?: string;
}

export interface WebhookSubscription {
  id: string;
  userId: string;
  url: string;
  secret: string;
  events: string[]; // e.g., ['remittance.*', 'payment.confirmed']
  active: boolean;
  createdAt: Date;
}

/**
 * Supported webhook events
 */
export const WEBHOOK_EVENTS = {
  // Payment events
  PAYMENT_PENDING: 'payment.pending',
  PAYMENT_CONFIRMED: 'payment.confirmed',
  PAYMENT_FAILED: 'payment.failed',
  
  // Conversion events
  CONVERSION_STARTED: 'conversion.started',
  CONVERSION_COMPLETED: 'conversion.completed',
  CONVERSION_FAILED: 'conversion.failed',
  
  // KYC events
  KYC_INITIATED: 'kyc.initiated',
  KYC_APPROVED: 'kyc.approved',
  KYC_REJECTED: 'kyc.rejected',
  
  // Account events
  ACCOUNT_VERIFYING: 'account.verifying',
  ACCOUNT_VERIFIED: 'account.verified',
  ACCOUNT_OPENING: 'account.opening',
  ACCOUNT_OPENED: 'account.opened',
  
  // Transfer events
  TRANSFER_INITIATED: 'transfer.initiated',
  TRANSFER_PROCESSING: 'transfer.processing',
  TRANSFER_COMPLETED: 'transfer.completed',
  TRANSFER_FAILED: 'transfer.failed',
  
  // Remittance events
  REMITTANCE_CREATED: 'remittance.created',
  REMITTANCE_COMPLETED: 'remittance.completed',
  REMITTANCE_FAILED: 'remittance.failed',
  REMITTANCE_CANCELLED: 'remittance.cancelled',
} as const;

/**
 * Create webhook event
 */
export async function createWebhookEvent(params: {
  remittanceId: string;
  event: string;
  data: Record<string, any>;
}): Promise<WebhookEvent> {
  const event: WebhookEvent = {
    id: `evt_${crypto.randomBytes(16).toString('hex')}`,
    remittanceId: params.remittanceId,
    event: params.event,
    data: params.data,
    timestamp: new Date(),
  };

  // Store in database
  // await db.createWebhookEvent(event);

  // Trigger delivery to all subscribed webhooks
  await deliverWebhookEvent(event);

  return event;
}

/**
 * Deliver webhook event to all subscribers
 */
async function deliverWebhookEvent(event: WebhookEvent): Promise<void> {
  // Get all active subscriptions that match this event
  const subscriptions = await getMatchingSubscriptions(event.event);

  // Create delivery records for each subscription
  const deliveries = subscriptions.map(sub => ({
    id: `del_${crypto.randomBytes(16).toString('hex')}`,
    webhookEventId: event.id,
    url: sub.url,
    status: 'pending' as const,
    attempts: 0,
  }));

  // Store deliveries in database
  // await db.createWebhookDeliveries(deliveries);

  // Attempt immediate delivery
  for (const delivery of deliveries) {
    const subscription = subscriptions.find(s => s.url === delivery.url);
    if (subscription) {
      await attemptWebhookDelivery(event, delivery, subscription);
    }
  }
}

/**
 * Attempt webhook delivery with retry logic
 */
export async function attemptWebhookDelivery(
  event: WebhookEvent,
  delivery: WebhookDelivery,
  subscription: WebhookSubscription
): Promise<WebhookDelivery> {
  delivery.attempts++;
  delivery.lastAttemptAt = new Date();

  try {
    // Generate signature
    const signature = generateWebhookSignature(event, subscription.secret);

    // Prepare payload
    const payload = {
      id: event.id,
      event: event.event,
      remittanceId: event.remittanceId,
      data: event.data,
      timestamp: event.timestamp.toISOString(),
    };

    // Send webhook
    const response = await fetch(subscription.url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Webhook-Signature': signature,
        'X-Webhook-Event': event.event,
        'X-Webhook-ID': event.id,
        'User-Agent': 'PaymentSwitch-Webhooks/1.0',
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(30000), // 30 second timeout
    });

    delivery.responseCode = response.status;

    if (response.ok) {
      delivery.status = 'delivered';
      delivery.responseBody = await response.text();
    } else {
      delivery.status = 'failed';
      delivery.error = `HTTP ${response.status}: ${response.statusText}`;
      delivery.responseBody = await response.text();

      // Schedule retry if not at max attempts
      if (delivery.attempts < 5) {
        delivery.nextRetryAt = calculateNextRetry(delivery.attempts);
      }
    }
  } catch (error) {
    delivery.status = 'failed';
    delivery.error = error instanceof Error ? error.message : 'Unknown error';

    // Schedule retry if not at max attempts
    if (delivery.attempts < 5) {
      delivery.nextRetryAt = calculateNextRetry(delivery.attempts);
    }
  }

  // Update delivery in database
  const db = await getDb();
  if (db && delivery.id) {
    try {
      await db.update(webhookDeliveries)
        .set({
          status: delivery.status,
          responseCode: delivery.responseCode ?? undefined,
          responseBody: delivery.responseBody ?? undefined,
          attempts: delivery.attempts,
          nextRetryAt: delivery.nextRetryAt ?? undefined,
          deliveredAt: delivery.status === 'delivered' ? new Date() : undefined,
        })
        .where(eq(webhookDeliveries.webhookId, delivery.id));
    } catch (err) {
      log.error({ err }, '[Webhook] Failed to update delivery status');
    }
  }

  return delivery;
}

/**
 * Generate HMAC-SHA256 signature for webhook
 */
function generateWebhookSignature(event: WebhookEvent, secret: string): string {
  const payload = JSON.stringify({
    id: event.id,
    event: event.event,
    remittanceId: event.remittanceId,
    data: event.data,
    timestamp: event.timestamp.toISOString(),
  });

  const hmac = crypto.createHmac('sha256', secret);
  hmac.update(payload);
  return hmac.digest('hex');
}

/**
 * Verify webhook signature
 */
export function verifyWebhookSignature(
  payload: string,
  signature: string,
  secret: string
): boolean {
  const hmac = crypto.createHmac('sha256', secret);
  hmac.update(payload);
  const expectedSignature = hmac.digest('hex');

  return crypto.timingSafeEqual(
    Buffer.from(signature),
    Buffer.from(expectedSignature)
  );
}

/**
 * Calculate next retry time with exponential backoff
 */
function calculateNextRetry(attempts: number): Date {
  // Retry schedule: 1m, 5m, 15m, 1h, 6h
  const delays = [60, 300, 900, 3600, 21600]; // in seconds
  const delay = delays[Math.min(attempts - 1, delays.length - 1)];
  
  return new Date(Date.now() + delay * 1000);
}

/**
 * Get subscriptions matching an event
 */
async function getMatchingSubscriptions(event: string): Promise<WebhookSubscription[]> {
  const db = await getDb();
  if (db) {
    try {
      const rows = await db.select().from(webhookDeliveries).where(eq(webhookDeliveries.event, event)).limit(10);
      return rows.map(r => ({
        id: `sub_${r.id}`,
        userId: '',
        url: r.url,
        secret: '',
        events: [r.event],
        active: true,
        createdAt: r.createdAt,
      }));
    } catch (err) {
      log.error({ err }, '[Webhook] DB subscriptions query error');
    }
  }
  return [];
}

/**
 * Check if event matches subscription patterns
 */
function matchesEventPattern(event: string, patterns: string[]): boolean {
  return patterns.some(pattern => {
    // Convert pattern to regex (e.g., 'remittance.*' -> /^remittance\..+$/)
    const regexPattern = pattern
      .replace(/\./g, '\\.')
      .replace(/\*/g, '.+');
    const regex = new RegExp(`^${regexPattern}$`);
    return regex.test(event);
  });
}

/**
 * Create webhook subscription
 */
export async function createWebhookSubscription(params: {
  userId: string;
  url: string;
  events: string[];
}): Promise<WebhookSubscription> {
  // Generate webhook secret
  const secret = crypto.randomBytes(32).toString('hex');

  const subscription: WebhookSubscription = {
    id: `sub_${crypto.randomBytes(16).toString('hex')}`,
    userId: params.userId,
    url: params.url,
    secret,
    events: params.events,
    active: true,
    createdAt: new Date(),
  };

  const db = await getDb();
  if (db) {
    try {
      await db.insert(webhookDeliveries).values({
        webhookId: subscription.id,
        event: params.events.join(','),
        url: params.url,
        status: 'active',
        attempts: 0,
      });
    } catch (err) {
      log.error({ err }, '[Webhook] DB persist error');
    }
  }

  return subscription;
}

/**
 * Update webhook subscription
 */
export async function updateWebhookSubscription(params: {
  subscriptionId: string;
  url?: string;
  events?: string[];
  active?: boolean;
}): Promise<WebhookSubscription | null> {
  log.info({ subscriptionId: params.subscriptionId }, '[Webhook] Update subscription');
  return null;
}

/**
 * Delete webhook subscription
 */
export async function deleteWebhookSubscription(
  subscriptionId: string
): Promise<boolean> {
  const db = await getDb();
  if (db) {
    try {
      await db.delete(webhookDeliveries).where(eq(webhookDeliveries.webhookId, subscriptionId));
    } catch (err) {
      log.error({ err }, '[Webhook] DB delete error');
    }
  }
  return true;
}

/**
 * Get webhook deliveries for an event
 */
export async function getWebhookDeliveries(
  eventId: string
): Promise<WebhookDelivery[]> {
  const db = await getDb();
  if (db) {
    try {
      const rows = await db.select().from(webhookDeliveries)
        .where(eq(webhookDeliveries.webhookId, eventId))
        .orderBy(desc(webhookDeliveries.createdAt));
      return rows.map(r => ({
        id: `del_${r.id}`,
        webhookEventId: eventId,
        subscriptionId: r.webhookId,
        url: r.url,
        status: r.status as 'pending' | 'delivered' | 'failed',
        statusCode: r.responseCode || undefined,
        responseBody: r.responseBody || undefined,
        attempts: r.attempts,
        nextRetryAt: r.nextRetryAt || undefined,
        deliveredAt: r.deliveredAt || undefined,
      }));
    } catch (err) {
      log.error({ err }, '[Webhook] DB deliveries query error');
    }
  }
  return [];
}

/**
 * Retry failed webhook delivery
 */
export async function retryWebhookDelivery(
  deliveryId: string
): Promise<WebhookDelivery | null> {
  log.info({ deliveryId }, '[Webhook] Retry delivery');
  return null;
}

/**
 * Process pending webhook deliveries (for background job)
 */
export async function processPendingWebhooks(): Promise<{
  processed: number;
  succeeded: number;
  failed: number;
}> {
  let processed = 0;
  let succeeded = 0;
  let failed = 0;

  const db = await getDb();
  if (db) {
    try {
      const pending = await db.select().from(webhookDeliveries)
        .where(and(
          eq(webhookDeliveries.status, 'pending'),
          lte(webhookDeliveries.nextRetryAt, new Date())
        ))
        .limit(100);

      for (const delivery of pending) {
        processed++;
        try {
          const payload = JSON.stringify(delivery.payload);
          const response = await fetch(delivery.url, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: payload,
          });
          if (response.ok) {
            await db.update(webhookDeliveries)
              .set({ status: 'delivered', responseCode: response.status, deliveredAt: new Date() })
              .where(eq(webhookDeliveries.id, delivery.id));
            succeeded++;
          } else {
            await db.update(webhookDeliveries)
              .set({
                status: delivery.attempts >= 5 ? 'failed' : 'pending',
                responseCode: response.status,
                attempts: delivery.attempts + 1,
                nextRetryAt: calculateNextRetry(delivery.attempts + 1),
              })
              .where(eq(webhookDeliveries.id, delivery.id));
            failed++;
          }
        } catch (err) {
          failed++;
        }
      }
    } catch (err) {
      log.error({ err }, '[Webhook] Process pending error');
    }
  }

  return { processed, succeeded, failed };
}

/**
 * Get webhook event by ID
 */
export async function getWebhookEvent(eventId: string): Promise<WebhookEvent | null> {
  const db = await getDb();
  if (db) {
    try {
      const rows = await db.select().from(webhookDeliveries).where(eq(webhookDeliveries.webhookId, eventId)).limit(1);
      if (rows.length > 0) {
        const r = rows[0];
        return {
          id: eventId,
          remittanceId: r.remittanceId || '',
          event: r.event,
          data: (r.payload as Record<string, any>) || {},
          timestamp: r.createdAt,
        };
      }
    } catch (err) {
      log.error({ err }, '[Webhook] DB event query error');
    }
  }
  return null;
}

/**
 * List webhook events for a remittance
 */
export async function listWebhookEvents(params: {
  remittanceId: string;
  limit?: number;
  offset?: number;
}): Promise<{
  events: WebhookEvent[];
  total: number;
}> {
  const db = await getDb();
  if (db) {
    try {
      const rows = await db.select().from(webhookDeliveries)
        .where(eq(webhookDeliveries.remittanceId, params.remittanceId))
        .orderBy(desc(webhookDeliveries.createdAt))
        .limit(params.limit || 20);
      return {
        events: rows.map(r => ({
          id: r.webhookId,
          remittanceId: r.remittanceId || '',
          event: r.event,
          data: (r.payload as Record<string, any>) || {},
          timestamp: r.createdAt,
        })),
        total: rows.length,
      };
    } catch (err) {
      log.error({ err }, '[Webhook] DB events list error');
    }
  }
  return { events: [], total: 0 };
}

/**
 * Test webhook endpoint
 */
export async function testWebhookEndpoint(params: {
  url: string;
  secret: string;
}): Promise<{
  success: boolean;
  responseCode?: number;
  responseTime?: number;
  error?: string;
}> {
  const startTime = Date.now();

  try {
    // Create test event
    const testEvent: WebhookEvent = {
      id: 'evt_test',
      remittanceId: 'rem_test',
      event: 'webhook.test',
      data: { message: 'This is a test webhook' },
      timestamp: new Date(),
    };

    // Generate signature
    const signature = generateWebhookSignature(testEvent, params.secret);

    // Send test webhook
    const response = await fetch(params.url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Webhook-Signature': signature,
        'X-Webhook-Event': testEvent.event,
        'X-Webhook-ID': testEvent.id,
      },
      body: JSON.stringify({
        id: testEvent.id,
        event: testEvent.event,
        remittanceId: testEvent.remittanceId,
        data: testEvent.data,
        timestamp: testEvent.timestamp.toISOString(),
      }),
      signal: AbortSignal.timeout(10000),
    });

    const responseTime = Date.now() - startTime;

    return {
      success: response.ok,
      responseCode: response.status,
      responseTime,
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error',
      responseTime: Date.now() - startTime,
    };
  }
}
